Jetstream can't restore after a network breakdown

I’m deploying a Jetstream cluster with 3 nodes in k8s.
It occurs that Jetstream doesn’t work after a network breakdown . 
The same issue occurred yesterday. And it can only resume after manual restarting the  servers.
![Uploading image.png…]()

@wchajl Looks like the 2 stacks that you provided show that the server is actively recovering the state since the routine reading blocks has no indication that it is blocked:
```
goroutine 1 [syscall]:
syscall.Syscall(0x0, 0x7, 0xc004734000, 0xfff12b)
	/home/travis/.gimme/versions/go1.17.8.linux.amd64/src/syscall/asm_linux_amd64.s:20 +0x5
syscall.read(0xc0003098c0, {0xc004734000, 0x0, 0x42c0a0})
	/home/travis/.gimme/versions/go1.17.8.linux.amd64/src/syscall/zsyscall_linux_amd64.go:687 +0x4d
syscall.Read(...)
	/home/travis/.gimme/versions/go1.17.8.linux.amd64/src/syscall/syscall_unix.go:189
internal/poll.ignoringEINTRIO(...)
	/home/travis/.gimme/versions/go1.17.8.linux.amd64/src/internal/poll/fd_unix.go:582
internal/poll.(*FD).Read(0xc0003098c0, {0xc004734000, 0xfff12b, 0xfff12b})
	/home/travis/.gimme/versions/go1.17.8.linux.amd64/src/internal/poll/fd_unix.go:163 +0x285
os.(*File).read(...)
	/home/travis/.gimme/versions/go1.17.8.linux.amd64/src/os/file_posix.go:32
os.(*File).Read(0xc0000b00f0, {0xc004734000, 0xc00016b9b8, 0x44c512})
	/home/travis/.gimme/versions/go1.17.8.linux.amd64/src/os/file.go:119 +0x5e
io.ReadAtLeast({0xaf18e0, 0xc0000b00f0}, {0xc004734000, 0xfff12b, 0xfff12b}, 0xfff12b)
	/home/travis/.gimme/versions/go1.17.8.linux.amd64/src/io/io.go:328 +0x9a
io.ReadFull(...)
	/home/travis/.gimme/versions/go1.17.8.linux.amd64/src/io/io.go:347
github.com/nats-io/nats-server/v2/server.(*msgBlock).loadBlock(0xc0000001a0, {0x0, 0x0, 0x0})
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/filestore.go:3236 +0x136
github.com/nats-io/nats-server/v2/server.(*msgBlock).loadMsgsWithLock(0xc001571860)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/filestore.go:3282 +0x2e9
github.com/nats-io/nats-server/v2/server.(*msgBlock).generatePerSubjectInfo(0xc001571860)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/filestore.go:4415 +0xef
github.com/nats-io/nats-server/v2/server.(*msgBlock).readPerSubjectInfo(0xc001571860)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/filestore.go:4467 +0x394
github.com/nats-io/nats-server/v2/server.(*fileStore).recoverMsgBlock(0xc000392280, {0xaffff8, 0xc00028bba0}, 0x13c5)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/filestore.go:611 +0xff0
github.com/nats-io/nats-server/v2/server.(*fileStore).recoverMsgs(0xc000392280)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/filestore.go:881 +0x858
github.com/nats-io/nats-server/v2/server.newFileStoreWithCreated({{0xc000228750, 0x25}, _, _, _, _}, {{0xc000233370, 0x9}, {0x0, 0x0}, ...}, ...)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/filestore.go:318 +0x774
github.com/nats-io/nats-server/v2/server.(*stream).setupStore(0xc000392000, 0xc00016ca90)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/stream.go:2459 +0x2d5
github.com/nats-io/nats-server/v2/server.(*Account).addStreamWithAssignment(0xc000127d40, 0xc000380018, 0x0, 0x0)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/stream.go:413 +0xdcb
github.com/nats-io/nats-server/v2/server.(*Account).addStream(0xc000233380, 0xc000232930)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/stream.go:264 +0x1d
github.com/nats-io/nats-server/v2/server.(*Account).EnableJetStream(0xc000127d40, 0xe0a3a0)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/jetstream.go:1159 +0x2be5
github.com/nats-io/nats-server/v2/server.(*Server).configJetStream(0xc0001c6000, 0xc000127d40)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/jetstream.go:630 +0x9c
github.com/nats-io/nats-server/v2/server.(*Server).configAllJetStreamAccounts(0xc0001c6000)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/jetstream.go:677 +0x176
github.com/nats-io/nats-server/v2/server.(*Server).enableJetStreamAccounts(0xc00024e370)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/jetstream.go:563 +0xb1
github.com/nats-io/nats-server/v2/server.(*Server).enableJetStream(0xc0001c6000, {0x40000000, 0x3200000000, {0xc000232450, 0xf}, {0xc00019e99d, 0xb}})
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/jetstream.go:377 +0x919
github.com/nats-io/nats-server/v2/server.(*Server).EnableJetStream(0xc0001c6000, 0xc00016de38)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/jetstream.go:189 +0x40c
github.com/nats-io/nats-server/v2/server.(*Server).Start(0xc0001c6000)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/server.go:1759 +0xf65
github.com/nats-io/nats-server/v2/server.Run(...)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/service.go:22
main.main()
	/home/travis/gopath/src/github.com/nats-io/nats-server/main.go:118 +0x2fa
```

Between stacksz.log and nats_0_stacksz.log (which is from the same server), I see a 2 minutes difference in other go routines, but that specific go routine does not show being stuck in one system call, hence my assumption that it is actively recovering.

The nats-1.log shows that it is isolated from the other servers, or that there are routes that are no longer reachable:
```
[12] 2022/03/22 11:21:59.255414 [DBG] Trying to connect to route on idms-native-test-nats-2.idms-native-test-nats.idms-nats.svc.cluster.local:6222 (100.101.145.66:6222)
[12] 2022/03/22 11:21:59.256652 [ERR] Error trying to connect to route (attempt 3923): dial tcp 100.101.145.66:6222: connect: connection refused
[12] 2022/03/22 11:21:59.340148 [DBG] Trying to connect to route on idms-native-test-nats-0.idms-native-test-nats.idms-nats.svc.cluster.local:6222 (100.101.138.72:6222)
[12] 2022/03/22 11:21:59.340604 [ERR] Error trying to connect to route (attempt 255): dial tcp 100.101.138.72:6222: connect: connection refused
```

Regarding the logs, nats0_previous.log, which shows that it started and is a 2.7.4 server, took more than 20 minutes to recover the state for about 10 million messages and 5 consumers:
```
[7869] 2022/03/22 10:55:30.843584 [DBG] Recovering JetStream state for account "CEC"
[7869] 2022/03/22 11:17:41.990194 [INF]   Restored 9,902,004 messages for stream 'CEC > cecstream'
[7869] 2022/03/22 11:17:41.990948 [INF]   Recovering 5 consumers for stream - 'CEC > cecstream'
```
It also shows the panic decoding the consumer state:
```
[7869] 2022/03/22 11:17:42.773338 [DBG] Starting consumer monitor for 'CEC > cecstream > eventPersister2' [C-R3F-YzqQD7aI]
(...)
[7869] 2022/03/22 11:17:42.804898 [ERR] JetStream cluster could not decode consumer snapshot for 'CEC > cecstream > eventPersister2' [C-R3F-YzqQD7aI]
[7869] 2022/03/22 11:17:42.804939 [DBG] Exiting consumer monitor for 'CEC > cecstream > eventPersister2' [C-R3F-YzqQD7aI]
panic: corrupt state file

goroutine 4603 [running]:
github.com/nats-io/nats-server/v2/server.(*jetStream).applyConsumerEntries(0xc0001306e0, 0xc000221000, 0x0, 0x0)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3167 +0x79c
github.com/nats-io/nats-server/v2/server.(*jetStream).monitorConsumer(0xc000174000, 0xc000221000, 0xc002226a20)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3134 +0x835
github.com/nats-io/nats-server/v2/server.(*jetStream).processClusterCreateConsumer.func1()
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:2923 +0x25
created by github.com/nats-io/nats-server/v2/server.(*Server).startGoRoutine
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/server.go:3024 +0x87
```
I will let @derekcollison comment on that on what could be done to skip that (maybe delete the corresponding consumer directory?)

Finally, you mentioned that you have deployed in 3 different k8s clusters, are you saying that you have 3 NATS clusters connected with gateways, or is it that you have 1 NATS cluster of 3 nodes, each node running in a k8s cluster? If the former, you need to make sure that each node has a unique server name! The server detects duplicate server names within a cluster, but does not across gateways (this will be the case in the next release).

--- a/locksordering.txt
+++ b/locksordering.txt
@@ -3,3 +3,5 @@ Here is the list of some established lock ordering.
 In this list, A -> B means that you can have A.Lock() then B.Lock(), not the opposite.
 
 jetStream -> jsAccount -> Server -> client-> Account
+
+stream -> consumer

--- a/server/consumer.go
+++ b/server/consumer.go
@@ -795,12 +795,12 @@ func (o *consumer) setLeader(isLeader bool) {
 		}
 
 		mset.mu.RLock()
-		s, jsa, stream := mset.srv, mset.jsa, mset.cfg.Name
+		s, jsa, stream, lseq := mset.srv, mset.jsa, mset.cfg.Name, mset.lseq
 		mset.mu.RUnlock()
 
 		o.mu.Lock()
 		// Restore our saved state. During non-leader status we just update our underlying store.
-		o.readStoredState()
+		o.readStoredState(lseq)
 
 		// Do info sub.
 		if o.infoSub == nil && jsa != nil {
@@ -1787,10 +1787,10 @@ func (o *consumer) ackWait(next time.Duration) time.Duration {
 }
 
 // Due to bug in calculation of sequences on restoring redelivered let's do quick sanity check.
-func (o *consumer) checkRedelivered() {
+func (o *consumer) checkRedelivered(slseq uint64) {
 	var lseq uint64
 	if mset := o.mset; mset != nil {
-		lseq = mset.lastSeq()
+		lseq = slseq
 	}
 	var shouldUpdateState bool
 	for sseq := range o.rdc {
@@ -1807,7 +1807,7 @@ func (o *consumer) checkRedelivered() {
 
 // This will restore the state from disk.
 // Lock should be held.
-func (o *consumer) readStoredState() error {
+func (o *consumer) readStoredState(slseq uint64) error {
 	if o.store == nil {
 		return nil
 	}
@@ -1815,7 +1815,7 @@ func (o *consumer) readStoredState() error {
 	if err == nil && state != nil && (state.Delivered.Consumer != 0 || state.Delivered.Stream != 0) {
 		o.applyState(state)
 		if len(o.rdc) > 0 {
-			o.checkRedelivered()
+			o.checkRedelivered(slseq)
 		}
 	}
 	return err
@@ -3570,14 +3570,14 @@ func (o *consumer) hasNoLocalInterest() bool {
 // This is when the underlying stream has been purged.
 // sseq is the new first seq for the stream after purge.
 // Lock should be held.
-func (o *consumer) purge(sseq uint64) {
+func (o *consumer) purge(sseq uint64, slseq uint64) {
 	// Do not update our state unless we know we are the leader.
 	if !o.isLeader() {
 		return
 	}
 	// Signals all have been purged for this consumer.
 	if sseq == 0 {
-		sseq = o.mset.lastSeq() + 1
+		sseq = slseq + 1
 	}
 
 	o.mu.Lock()
@@ -3713,7 +3713,7 @@ func (o *consumer) stopWithFlags(dflag, sdflag, doSignal, advisory bool) error {
 		stop := mset.lastSeq()
 		o.mu.Lock()
 		if !o.isLeader() {
-			o.readStoredState()
+			o.readStoredState(stop)
 		}
 		start := o.asflr
 		o.mu.Unlock()

--- a/server/consumer.go
+++ b/server/consumer.go
@@ -2775,15 +2775,24 @@ func (o *consumer) loopAndGatherMsgs(qch chan struct{}) {
 		lseq uint64
 	)
 
+	o.mu.RLock()
+	mset := o.mset
+	getLSeq := o.replay
+	o.mu.RUnlock()
+	// consumer is closed when mset is set to nil.
+	if mset == nil {
+		return
+	}
+	if getLSeq {
+		lseq = mset.state().LastSeq
+	}
+
 	o.mu.Lock()
 	s := o.srv
-	if o.replay {
-		// consumer is closed when mset is set to nil.
-		if o.mset == nil {
-			o.mu.Unlock()
-			return
-		}
-		lseq = o.mset.state().LastSeq
+	// need to check again if consumer is closed
+	if o.mset == nil {
+		o.mu.Unlock()
+		return
 	}
 	// For idle heartbeat support.
 	var hbc <-chan time.Time

--- a/server/jetstream.go
+++ b/server/jetstream.go
@@ -1227,8 +1227,9 @@ func (a *Account) EnableJetStream(limits *JetStreamAccountLimits) error {
 			if !cfg.Created.IsZero() {
 				obs.setCreatedTime(cfg.Created)
 			}
+			lseq := e.mset.lastSeq()
 			obs.mu.Lock()
-			err = obs.readStoredState()
+			err = obs.readStoredState(lseq)
 			obs.mu.Unlock()
 			if err != nil {
 				s.Warnf("    Error restoring consumer state: %v", err)

--- a/server/jetstream_cluster_test.go
+++ b/server/jetstream_cluster_test.go
@@ -26,6 +26,7 @@ import (
 	"os"
 	"path/filepath"
 	"reflect"
+	"runtime"
 	"strconv"
 	"strings"
 	"sync"
@@ -11378,6 +11379,95 @@ func TestJetStreamDuplicateMsgIdsOnCatchupAndLeaderTakeover(t *testing.T) {
 	require_NoError(t, err)
 	if !pa.Duplicate {
 		t.Fatalf("Expected msg to be a duplicate")
+	}
+}
+
+func TestJetStreamClusterConsumerLeaderChangeDeadlock(t *testing.T) {
+	c := createJetStreamClusterExplicit(t, "R3S", 3)
+	defer c.shutdown()
+
+	nc, js := jsClientConnect(t, c.randomServer())
+	defer nc.Close()
+
+	// Create a stream and durable with ack explicit
+	_, err := js.AddStream(&nats.StreamConfig{Name: "test", Subjects: []string{"foo"}, Replicas: 3})
+	require_NoError(t, err)
+	_, err = js.AddConsumer("test", &nats.ConsumerConfig{
+		Durable:        "test",
+		DeliverSubject: "bar",
+		AckPolicy:      nats.AckExplicitPolicy,
+		AckWait:        250 * time.Millisecond,
+	})
+	require_NoError(t, err)
+
+	// Wait for a leader
+	c.waitOnConsumerLeader("$G", "test", "test")
+	cl := c.consumerLeader("$G", "test", "test")
+
+	// Publish a message
+	_, err = js.Publish("foo", []byte("msg"))
+	require_NoError(t, err)
+
+	// Create nats consumer on "bar" and don't ack it
+	sub := natsSubSync(t, nc, "bar")
+	natsNexMsg(t, sub, time.Second)
+	// Wait for redeliveries, to make sure it is in the redelivery map
+	natsNexMsg(t, sub, time.Second)
+	natsNexMsg(t, sub, time.Second)
+
+	mset, err := cl.GlobalAccount().lookupStream("test")
+	require_NoError(t, err)
+	require_True(t, mset != nil)
+
+	// There are parts in the code (for instance when signaling to consumers
+	// that there are new messages) where we get the mset lock and iterate
+	// over the consumers and get consumer lock. We are going to do that
+	// in a go routine while we send a consumer step down request from
+	// another go routine. We will watch for possible deadlock and if
+	// found report it.
+	ch := make(chan struct{})
+	doneCh := make(chan struct{})
+	go func() {
+		defer close(doneCh)
+		for {
+			mset.mu.Lock()
+			for _, o := range mset.consumers {
+				o.mu.Lock()
+				time.Sleep(time.Duration(rand.Intn(10)) * time.Millisecond)
+				o.mu.Unlock()
+			}
+			mset.mu.Unlock()
+			select {
+			case <-ch:
+				return
+			default:
+			}
+		}
+	}()
+
+	// Now cause a leader changes
+	for i := 0; i < 5; i++ {
+		m, err := nc.Request("$JS.API.CONSUMER.LEADER.STEPDOWN.test.test", nil, 2*time.Second)
+		// Ignore error here and check for deadlock below
+		if err != nil {
+			break
+		}
+		// if there is a message, check that it is success
+		var resp JSApiConsumerLeaderStepDownResponse
+		err = json.Unmarshal(m.Data, &resp)
+		require_NoError(t, err)
+		require_True(t, resp.Success)
+		c.waitOnConsumerLeader("$G", "test", "test")
+	}
+
+	close(ch)
+	select {
+	case <-doneCh:
+		// OK!
+	case <-time.After(2 * time.Second):
+		buf := make([]byte, 1000000)
+		n := runtime.Stack(buf, true)
+		t.Fatalf("Suspected deadlock, printing current stack. The test suite may timeout and will also dump the stack\n%s\n", buf[:n])
 	}
 }
 

--- a/server/stream.go
+++ b/server/stream.go
@@ -1147,6 +1147,7 @@ func (mset *stream) purge(preq *JSApiStreamPurgeRequest) (purged uint64, err err
 	var state StreamState
 	mset.store.FastState(&state)
 	fseq := state.FirstSeq
+	lseq := state.LastSeq
 
 	// Check for filtered purge.
 	if preq != nil && preq.Subject != _EMPTY_ {
@@ -1155,7 +1156,7 @@ func (mset *stream) purge(preq *JSApiStreamPurgeRequest) (purged uint64, err err
 	}
 
 	for _, o := range obs {
-		o.purge(fseq)
+		o.purge(fseq, lseq)
 	}
 	return purged, nil
 }
@@ -3811,6 +3812,7 @@ func (a *Account) RestoreStream(ncfg *StreamConfig, r io.Reader) (*stream, error
 	if !fcfg.Created.IsZero() {
 		mset.setCreatedTime(fcfg.Created)
 	}
+	lseq := mset.lastSeq()
 
 	// Now do consumers.
 	odir := filepath.Join(ndir, consumerDir)
@@ -3854,7 +3856,7 @@ func (a *Account) RestoreStream(ncfg *StreamConfig, r io.Reader) (*stream, error
 			obs.setCreatedTime(cfg.Created)
 		}
 		obs.mu.Lock()
-		err = obs.readStoredState()
+		err = obs.readStoredState(lseq)
 		obs.mu.Unlock()
 		if err != nil {
 			mset.stop(true, false)

